README.md
# meezer
Supervised Siamese-ish networks using hard negative examples.

![](https://i.redd.it/fq6l78zdlow21.jpg)

A good deal of this code, especially files dealing with building and extracting information from the `Annoy` index, is based off of [Ivis](https://github.com/beringresearch/ivis). The `Img2Vec` code is based off of [this repo](https://github.com/jaredwinick/img2vec-keras).

## Usage
```python
import matplotlib.pyplot as plt
import numpy as np
from sklearn.datasets import load_digits
from sklearn.preprocessing import MinMaxScaler

from meezer import Meezer


# GET DATA
digits = load_digits()

X, labels = digits.data, digits.target
labels = labels.astype(int)

# SCALE DATA
X_scaled = MinMaxScaler().fit_transform(X)

# TRAIN A MODEL
model = Meezer(embedding_dims=2,
               k=25,
               batch_size=512,
               epochs=30,
               sub_epochs=10)
embeddings = model.fit_transform(X=X_scaled, Y=labels)

# VISUALIZE EMBEDDINGS
for i, mnist_num in enumerate(set(labels)):
    mnist_num_idxs = np.argwhere(labels == mnist_num)
    plt.scatter(embeddings[mnist_num_idxs, 0],
                embeddings[mnist_num_idxs, 1],
                color=plt.cm.gist_rainbow(i / len(set(labels))),
                s=4,
                label=mnist_num)

plt.xlabel('embedding dim 1')
plt.ylabel('embedding dim 2')
plt.legend(bbox_to_anchor=(1.04,1), loc='upper left')

plt.show()
```

### Development
Begin by installing [Docker](https://docs.docker.com/install/) if you have not already. Once Docker is running, run development from within the Docker container:

```bash
# build the Docker image
docker build -t meezer .

# run the Docker container in interactive mode.
docker run \
    -it \
    --rm \
    -v "${PWD}:/meezer" \
    -p 8888:8888 \
    meezer /bin/bash

# then run JupyterLab and begin development
jupyter lab --ip 0.0.0.0 --no-browser --allow-root --NotebookApp.token='' --NotebookApp.password=''
```

## TODOs
- [ ] Exponential data `hard_mode` scheduler to introduce hard negative examples at a better rate than linearly.
- [ ] Clean up notebooks to show examples a bit better and with more available data sources.
- [ ] Add in traditional Siamese network using the triplet generator, but now feeding inputs to models two at a time, with `y` being a label of `1` if the pair are the same class or `0` otherwise. Using hard negative examples will help performance, most likely.
- [ ] Better support for callbacks (early stopping, learning rate scheduler, etc.).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project uses [Semantic Versioning](http://semver.org/).


# [0.2.1] - 2020-6-22
### Added
 - offset to `hard_mode`
 - Jupyter notebook examples
### Changed
 - `README` example simplification

# [0.2.0] - 2020-6-22
### Added
 - `Dockerfile` now adds `conda` and `Jupyter` setup
 - `README` example
 - `Img2Vec` helper class
### Fixed
 - `knn_triplet_from_neighbor_list` now checks negative hard examples are not in any of the labelled examples

# [0.1.1] - 2020-6-18
### Added
 - `Dockerfile`
 - better docs

# [0.1.0] - 2020-6-18
### Added
 - working prototype for `Meezer` pipeline

# [0.0.0] - 2020-6-18
### Added
 - initial project scaffolding from generator-python-repo


meezer/knn.py
from collections import namedtuple
from multiprocessing import cpu_count, Process, Queue
from operator import attrgetter
import time

from annoy import AnnoyIndex
import numpy as np
from scipy.sparse import issparse
from tqdm.auto import tqdm


def build_annoy_index(X, path, ntrees=50, build_index_on_disk=True, verbose=True):
    """
    Build a standalone Annoy index.

    Parameters
    -------------
    X: np.array with shape (n_samples, n_features)
    path: str or Path
        The filepath of a trained annoy index file saved on disk
    ntrees: int
        The number of random projections trees built by Annoy to approximate KNN. The more trees,the
        higher the memory usage, but the better the accuracy of results (default 50)
    build_index_on_disk: bool
        Whether to build the annoy index directly on disk. Building on disk should allow for bigger
        datasets to be indexed, but may cause issues. If None, on-disk building will be enabled for
        Linux, but not Windows due to issues on Windows.
    verbose: bool

    """
    verbose = int(verbose)

    index = AnnoyIndex(X.shape[1], metric='angular')
    if build_index_on_disk:
        index.on_disk_build(str(path))

    if issparse(X):
        for i in tqdm(range(X.shape[0]), disable=verbose < 1):
            v = X[i].toarray()[0]
            index.add_item(i, v)
    else:
        for i in tqdm(range(X.shape[0]), disable=verbose < 1):
            v = X[i]
            index.add_item(i, v)

    try:
        index.build(ntrees)
    except Exception:
        raise IndexBuildingError(
            'Error building Annoy Index. Try setting `build_index_on_disk` to False.'
        )
    else:
        if not build_index_on_disk:
            index.save(path)
        return index


def extract_knn(X, index_filepath, k=150, search_k=-1, verbose=1):
    """
    Starts multiple processes to retrieve nearest neighbors using an Annoy Index
    in parallel.

    """
    n_dims = X.shape[1]
    index_filepath = str(index_filepath)

    chunk_size = X.shape[0] // cpu_count()
    remainder = (X.shape[0] % cpu_count()) > 0
    process_pool = []
    results_queue = Queue()

    # Split up the indices and assign processes for each chunk
    i = 0
    while (i + chunk_size) <= X.shape[0]:
        process_pool.append(
            KNN_Worker(
                index_filepath, k, search_k, n_dims, (i, i + chunk_size), results_queue
            )
        )
        i += chunk_size
    if remainder:
        process_pool.append(
            KNN_Worker(
                index_filepath, k, search_k, n_dims, (i, X.shape[0]), results_queue
            )
        )

    try:
        for process in process_pool:
            process.start()

        # Read from queue constantly to prevent it from becoming full
        with tqdm(total=X.shape[0], disable=verbose < 1) as pbar:
            neighbor_list = []
            neighbor_list_length = len(neighbor_list)
            while any(process.is_alive() for process in process_pool):
                while not results_queue.empty():
                    neighbor_list.append(results_queue.get())
                progress = len(neighbor_list) - neighbor_list_length
                pbar.update(progress)
                neighbor_list_length = len(neighbor_list)
                time.sleep(0.1)

            while not results_queue.empty():
                neighbor_list.append(results_queue.get())

        neighbor_list = sorted(neighbor_list, key=attrgetter('row_index'))
        neighbor_list = list(map(attrgetter('neighbor_list'), neighbor_list))

        return np.array(neighbor_list)
    except Exception:
        print('Halting KNN retrieval and cleaning up.')
        for process in process_pool:
            process.terminate()
        raise


IndexNeighbors = namedtuple('IndexNeighbors', 'row_index neighbor_list')


class KNN_Worker(Process):
    """
    Upon construction, this worker process loads an annoy index from disk.
    When started, the neighbors of the data-points specified by `data_indices`
    will be retrieved from the index according to the provided parameters
    and stored in the 'results_queue'.

    `data_indices` is a tuple of integers denoting the start and end range of
    indices to retrieve.

    """
    def __init__(
        self, index_filepath, k, search_k, n_dims, data_indices, results_queue
    ):
        self.index_filepath = index_filepath
        self.k = k
        self.n_dims = n_dims
        self.search_k = search_k
        self.data_indices = data_indices
        self.results_queue = results_queue
        super(KNN_Worker, self).__init__()

    def run(self):
        """TODO."""
        try:
            index = AnnoyIndex(self.n_dims, metric='angular')
            index.load(self.index_filepath)
            for i in range(self.data_indices[0], self.data_indices[1]):
                neighbor_indexes = index.get_nns_by_item(
                    i, self.k, search_k=self.search_k, include_distances=False
                )
                neighbor_indexes = np.array(neighbor_indexes, dtype=np.uint32)
                self.results_queue.put(
                    IndexNeighbors(row_index=i, neighbor_list=neighbor_indexes)
                )
        except Exception as e:
            self.exception = e
        finally:
            self.results_queue.close()


class IndexBuildingError(OSError):
    """TODO."""
    pass


meezer/losses.py
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import backend as K


def triplet_loss(distance='pn', margin=1):
    losses = get_loss_functions(margin=margin)

    loss_function = losses[distance]
    loss_function.__name__ = distance

    return loss_function


def get_loss_functions(margin=1):
    losses = {
        'pn': pn_loss(margin=margin),
        'euclidean': euclidean_loss(margin=margin),
        'softmax_ratio': softmax_ratio,
        'softmax_ratio_pn': softmax_ratio_pn,
        'manhattan': manhattan_loss(margin=margin),
        'manhattan_pn': manhattan_pn_loss(margin=margin),
        'chebyshev': chebyshev_loss(margin=margin),
        'chebyshev_pn': chebyshev_pn_loss(margin=margin),
        'cosine': cosine_loss(margin=margin),
        'cosine_pn': cosine_pn_loss(margin=margin),
    }
    return losses


def semi_supervised_loss(loss_function):
    def new_loss_function(y_true, y_pred):
        mask = tf.cast(~tf.math.equal(y_true, -1), tf.float32)
        y_true_pos = tf.nn.relu(y_true)
        loss = loss_function(y_true_pos, y_pred)
        masked_loss = loss * mask
        return masked_loss

    new_func = new_loss_function
    new_func.__name__ = loss_function.__name__
    return new_func


def is_hinge(supervised_loss):
    loss = keras.losses.get(supervised_loss)
    if loss in get_hinge_losses():
        return True
    return False


def get_hinge_losses():
    hinge_losses = set(
        [keras.losses.hinge, keras.losses.squared_hinge, keras.losses.categorical_hinge]
    )
    return hinge_losses


def get_categorical_losses():
    categorical_losses = set(
        [
            keras.losses.sparse_categorical_crossentropy,
            keras.losses.categorical_crossentropy,
            keras.losses.categorical_hinge,
            keras.losses.binary_crossentropy,
            keras.losses.hinge,
            keras.losses.squared_hinge,
        ]
    )
    return categorical_losses


def get_multiclass_losses():
    multiclass_losses = set(
        [
            keras.losses.sparse_categorical_crossentropy,
            keras.losses.categorical_crossentropy,
            keras.losses.categorical_hinge,
        ]
    )
    return multiclass_losses


def is_multiclass(supervised_loss):
    loss = keras.losses.get(supervised_loss)
    if loss in get_multiclass_losses():
        return True
    return False


def _euclidean_distance(x, y):
    return K.sqrt(
        K.maximum(K.sum(K.square(x - y), axis=-1, keepdims=True), K.epsilon())
    )


def _manhattan_distance(x, y):
    return K.sum(K.abs(x - y), axis=-1, keepdims=True)


def _chebyshev_distance(x, y):
    return K.max(K.abs(x - y), axis=-1, keepdims=True)


def _cosine_distance(x, y):
    return tf.losses.cosine_distance(
        tf.nn.l2_normalize(x, 0), tf.nn.l2_normalize(y, 0), axis=0
    )


def pn_loss(margin=1):
    def _pn_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)

        anchor_positive_distance = _euclidean_distance(anchor, positive)
        anchor_negative_distance = _euclidean_distance(anchor, negative)
        positive_negative_distance = _euclidean_distance(positive, negative)

        minimum_distance = K.min(
            K.concatenate([anchor_negative_distance, positive_negative_distance]),
            axis=-1,
            keepdims=True,
        )

        return K.mean(
            K.maximum(anchor_positive_distance - minimum_distance + margin, 0)
        )

    return _pn_loss


def euclidean_loss(margin=1):
    def _euclidean_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)
        return K.mean(
            K.maximum(
                _euclidean_distance(anchor, positive)
                - _euclidean_distance(anchor, negative)
                + margin,
                0,
            )
        )

    return _euclidean_loss


def manhattan_loss(margin=1):
    def _manhattan_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)
        return K.mean(
            K.maximum(
                _manhattan_distance(anchor, positive)
                - _manhattan_distance(anchor, negative)
                + margin,
                0,
            )
        )

    return _manhattan_loss


def manhattan_pn_loss(margin=1):
    def _pn_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)

        anchor_positive_distance = _manhattan_distance(anchor, positive)
        anchor_negative_distance = _manhattan_distance(anchor, negative)
        positive_negative_distance = _manhattan_distance(positive, negative)

        minimum_distance = K.min(
            K.concatenate([anchor_negative_distance, positive_negative_distance]),
            axis=-1,
            keepdims=True,
        )

        return K.mean(
            K.maximum(anchor_positive_distance - minimum_distance + margin, 0)
        )

    return _pn_loss


def chebyshev_loss(margin=1):
    def _chebyshev_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)
        return K.mean(
            K.maximum(
                _chebyshev_distance(anchor, positive)
                - _chebyshev_distance(anchor, negative)
                + margin,
                0,
            )
        )

    return _chebyshev_loss


def chebyshev_pn_loss(margin=1):
    def _pn_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)

        anchor_positive_distance = _chebyshev_distance(anchor, positive)
        anchor_negative_distance = _chebyshev_distance(anchor, negative)
        positive_negative_distance = _chebyshev_distance(positive, negative)

        minimum_distance = K.min(
            K.concatenate([anchor_negative_distance, positive_negative_distance]),
            axis=-1,
            keepdims=True,
        )

        return K.mean(
            K.maximum(anchor_positive_distance - minimum_distance + margin, 0)
        )

    return _pn_loss


def cosine_loss(margin=1):
    def _cosine_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)
        return K.mean(
            K.maximum(
                _cosine_distance(anchor, positive)
                - _cosine_distance(anchor, negative)
                + margin,
                0,
            )
        )

    return _cosine_loss


def cosine_pn_loss(margin=1):
    def _pn_loss(y_true, y_pred):
        anchor, positive, negative = tf.unstack(y_pred)

        anchor_positive_distance = _cosine_distance(anchor, positive)
        anchor_negative_distance = _cosine_distance(anchor, negative)
        positive_negative_distance = _cosine_distance(positive, negative)

        minimum_distance = K.min(
            tf.stack([anchor_negative_distance, positive_negative_distance]),
            axis=0,
            keepdims=True,
        )

        return K.mean(
            K.maximum(anchor_positive_distance - minimum_distance + margin, 0)
        )

    return _pn_loss


def softmax_ratio(y_true, y_pred):
    anchor, positive, negative = tf.unstack(y_pred)

    positive_distance = _euclidean_distance(anchor, positive)
    negative_distance = _euclidean_distance(anchor, negative)

    softmax = K.softmax(K.concatenate([positive_distance, negative_distance]))
    ideal_distance = K.variable([0, 1])
    return K.mean(K.maximum(softmax - ideal_distance, 0))


def softmax_ratio_pn(y_true, y_pred):
    anchor, positive, negative = tf.unstack(y_pred)

    anchor_positive_distance = _euclidean_distance(anchor, positive)
    anchor_negative_distance = _euclidean_distance(anchor, negative)
    positive_negative_distance = _euclidean_distance(positive, negative)

    minimum_distance = K.min(
        K.concatenate([anchor_negative_distance, positive_negative_distance]),
        axis=-1,
        keepdims=True,
    )

    softmax = K.softmax(K.concatenate([anchor_positive_distance, minimum_distance]))
    ideal_distance = K.variable([0, 1])
    return K.mean(K.maximum(softmax - ideal_distance, 0))


def validate_sparse_labels(Y):
    if not zero_indexed(Y):
        raise ValueError('Ensure that your labels are zero-indexed')
    if not consecutive_indexed(Y):
        raise ValueError('Ensure that your labels are indexed consecutively')


def zero_indexed(Y):
    if min(abs(Y)) != 0:
        return False
    return True


def consecutive_indexed(Y):
    """Assumes that `Y` is zero-indexed."""
    n_classes = len(np.unique(Y[Y != np.array(-1)]))
    if max(Y) >= n_classes:
        return False
    return True


meezer/meezer.py
import json
import os
import platform
import shutil

import numpy as np
from sklearn.base import BaseEstimator
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import backend as K
from tensorflow.keras import regularizers
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras.models import load_model, Model
from tqdm.auto import tqdm

from meezer.knn import build_annoy_index, extract_knn
from meezer.losses import (is_hinge,
                           is_multiclass,
                           semi_supervised_loss,
                           triplet_loss,
                           validate_sparse_labels)
from meezer.network import base_network, triplet_network
from meezer.triplet_generators import generator_from_index


class Meezer(BaseEstimator):
    """
    Meezer is a supervised Siamese network that uses a triplet loss function to seperate points from
    the same class and all other points. Over time, the pipeline introduces harder and harder
    negative examples to be used in the triplet loss using nearest-neighbors from an Annoy index
    generated after each epoch.

    Parameters
    -------------
    embedding_dims: int
        Number of dimensions in the embedding space (default 2)
    k: int
        The number of neighbors to retrieve for each point. Must be less than one minus the number
        of rows in the dataset (default 150)
    distance: str
        The loss function used to train the neural network. One of:
            * 'pn'
            * 'euclidean'
            * 'manhattan_pn'
            * 'manhattan'
            * 'chebyshev'
            * 'chebyshev_pn'
            * 'softmax_ratio_pn'
            * 'softmax_ratio'
            * 'cosine'
            * 'cosine_p'
        (default 'pn')
    batch_size: int
        The size of mini-batches used during gradient descent while training the neural network.
        Must be less than the number of rows in the dataset (default '128')
    epochs: int
        The maximum number of epochs to train the model for. Each epoch the network will see a
        triplet based on each data-point `sub_epochs` times. After each epoch, callbacks are reset
        and new hard negative examples are computed in an Annoy index (default 20)
    sub_epochs: int
        Number of epochs to train the model with a single set of hard negative examples for each of
        the epochs in `epochs` (default 10)
    margin: float
        The distance that is enforced between points by the triplet loss functions (default 1.)
    ntrees: int
        The number of random projections trees built by Annoy to approximate KNN. The more trees,
        the higher the memory usage, but the better the accuracy of results (default 50)
    search_k: int
        The maximum number of nodes inspected during a nearest neighbor query by Annoy. The higher,
        the more computation time required, but the higher the accuracy. The default is `-1`, which
        sets `search_k = n_trees * k`, where k is the number of neighbors to retrieve. If this is
        set too low, a variable number of neighbors may be retrieved per data-point (default -1)
    model: str or keras.models.Model
        The keras model to train using triplet loss. If a model object is provided, an embedding
        layer of size 'embedding_dims' will be appended to the end of the network. If a string, a
        pre-defined network by that name will be used. Possible options are:
            * 'szubert'
            * 'hinton'
            * 'maaten'
        (default 'szubert')
    supervision_metric: str or function
        The supervision metric to optimize when training keras in supervised mode. Supports all of
        the classification losses included with keras, so long as the labels are provided in the
        correct format. A list of keras' loss functions can be found at https://keras.io/losses/
        (default 'sparse_categorical_crossentropy')
    supervision_weight: float
        Float between 0 and 1 denoting the weighting to give to classification vs triplet loss when
        training in supervised mode. The higher the weight, the more classification influences
        training (default 0.25)
    annoy_index_path: str or Path
        The filepath of a pre-trained annoy index file to be saved on disk (default 'annoy.index')
    callbacks: list[keras.callbacks.Callback]
        List of keras Callbacks to pass model during training, such as the TensorBoard callback.
        Note these only apply to `sub_epochs` in each `epoch` before hard negative examples are
        re-computed (default [])
    early_stopping: bool
        Stop training the model if loss does not increase in `sub_epochs - 1` in any epoch in
        `epochs` (default True)
    reduce_lr_amount: float
        Amount to reduce learning rate by after each completed full epoch (default 1.)
    verbose: bool
        Controls the volume of logging output the model produces when training. When set to `False`,
        silences outputs, else will print outputs (default True)

    """
    def __init__(
        self,
        embedding_dims=2,
        k=150,
        distance='pn',
        batch_size=128,
        epochs=20,
        sub_epochs=10,
        margin=1.,
        ntrees=50,
        search_k=-1,
        model='szubert',
        supervision_metric='sparse_categorical_crossentropy',
        supervision_weight=0.25,
        annoy_index_path='annoy.index',
        callbacks=[],
        early_stopping=True,
        reduce_lr_amount=1.,
        verbose=True,
    ):
        self.embedding_dims = embedding_dims
        self.k = k
        self.distance = distance
        self.batch_size = batch_size
        self.epochs = epochs
        self.sub_epochs = sub_epochs
        self.margin = margin
        self.ntrees = ntrees
        self.search_k = search_k
        self.model_def = model
        self.supervision_metric = supervision_metric
        self.supervision_weight = supervision_weight
        self.annoy_index_path = annoy_index_path
        self.callbacks = callbacks
        self.early_stopping = early_stopping
        self.reduce_lr_amount = reduce_lr_amount
        self.verbose = int(verbose)

        self.model_ = None
        self.encoder = None
        self.supervised_model_ = None
        self.loss_history_ = list()

        self.hard_mode = True

        self.build_index_on_disk = True if platform.system() != 'Windows' else False

    def __getstate__(self):
        """Return object serializable variable dict."""
        state = dict(self.__dict__)
        if 'model_' in state:
            state['model_'] = None
        if 'encoder' in state:
            state['encoder'] = None
        if 'supervised_model_' in state:
            state['supervised_model_'] = None
        if 'callbacks' in state:
            state['callbacks'] = []
        if not isinstance(state['model_def'], str):
            state['model_def'] = None
        if 'neighbor_matrix' in state:
            state['neighbor_matrix'] = None
        return state

    def _fit(self, X, Y=None, shuffle_mode=True):
        if self.verbose > 0:
            print('Building KNN index...')
        build_annoy_index(
            X,
            self.annoy_index_path,
            ntrees=self.ntrees,
            build_index_on_disk=self.build_index_on_disk,
            verbose=self.verbose,
        )

        # sets up the triplet data
        datagen = generator_from_index(
            X,
            Y,
            index_path=self.annoy_index_path,
            k=self.k,
            batch_size=self.batch_size,
            search_k=self.search_k,
            verbose=self.verbose,
        )

        # setting up siamese network component
        loss_monitor = 'loss'
        try:
            triplet_loss_func = triplet_loss(distance=self.distance, margin=self.margin)
        except KeyError:
            raise ValueError(
                'Loss function `{}` not implemented.'.format(self.distance)
            )

        if type(self.model_def) is str:
            input_size = (X.shape[-1],)
            self.model_, anchor_embedding, _, _ = triplet_network(
                base_network(self.model_def, input_size),
                embedding_dims=self.embedding_dims,
            )
        else:
            self.model_, anchor_embedding, _, _ = triplet_network(
                self.model_def, embedding_dims=self.embedding_dims
            )

        # setting up supervised model component
        if not is_multiclass(self.supervision_metric):
            if not is_hinge(self.supervision_metric):
                # Binary logistic classifier
                if len(Y.shape) > 1:
                    self.n_classes = Y.shape[-1]
                else:
                    self.n_classes = 1
                supervised_output = Dense(
                    self.n_classes, activation='sigmoid', name='supervised'
                )(anchor_embedding)
            else:
                # Binary Linear SVM output
                if len(Y.shape) > 1:
                    self.n_classes = Y.shape[-1]
                else:
                    self.n_classes = 1
                supervised_output = Dense(
                    self.n_classes,
                    activation='linear',
                    name='supervised',
                    kernel_regularizer=regularizers.l2(),
                )(anchor_embedding)
        else:
            if not is_hinge(self.supervision_metric):
                validate_sparse_labels(Y)
                self.n_classes = len(np.unique(Y[Y != np.array(-1)]))
                # Softmax classifier
                supervised_output = Dense(
                    self.n_classes, activation='softmax', name='supervised'
                )(anchor_embedding)
            else:
                self.n_classes = len(np.unique(Y, axis=0))
                # Multiclass Linear SVM output
                supervised_output = Dense(
                    self.n_classes,
                    activation='linear',
                    name='supervised',
                    kernel_regularizer=regularizers.l2(),
                )(anchor_embedding)

        supervised_loss = keras.losses.get(self.supervision_metric)
        if self.supervision_metric == 'sparse_categorical_crossentropy':
            supervised_loss = semi_supervised_loss(supervised_loss)

        final_network = Model(
            inputs=self.model_.inputs, outputs=[self.model_.output, supervised_output]
        )
        self.model_ = final_network
        self.model_.compile(
            optimizer='adam',
            loss={'stacked_triplets': triplet_loss_func, 'supervised': supervised_loss},
            loss_weights={
                'stacked_triplets': 1 - self.supervision_weight,
                'supervised': self.supervision_weight,
            },
        )

        # Store dedicated classification model
        supervised_model_input = Input(shape=(X.shape[-1],))
        embedding = self.model_.layers[3](supervised_model_input)
        softmax_out = self.model_.layers[-1](embedding)

        self.supervised_model_ = Model(supervised_model_input, softmax_out)

        self.encoder = self.model_.layers[3]

        if self.verbose > 0:
            print('Training neural network...')

        if self.early_stopping:
            early_stopping_monitor = EarlyStopping(
                monitor=loss_monitor, patience=self.sub_epochs - 1,
            )
            self.callbacks.append(early_stopping_monitor)

        for epoch in tqdm(range(0, self.epochs)):
            if epoch > 0 and self.supervision_weight < 1:
                data_to_annoy = self.transform(X, verbose=False)

                if self.verbose:
                    print('Building new KNN index...')
                build_annoy_index(
                    data_to_annoy,
                    self.annoy_index_path,
                    ntrees=self.ntrees,
                    build_index_on_disk=self.build_index_on_disk,
                    verbose=0,
                )

                if self.verbose:
                    print('Extracting KNN from new index...')
                neighbor_matrix = extract_knn(
                    X=data_to_annoy,
                    index_filepath=self.annoy_index_path,
                    k=self.k,
                    search_k=self.search_k,
                    verbose=0
                )

                assert X.shape[0] == neighbor_matrix.shape[0]
                datagen.neighbor_matrix = neighbor_matrix
                # TODO: add in some exponential `hard_mode` scheduler
                datagen.hard_mode = (epoch / self.epochs) + 0.2 if self.hard_mode is True else 0

                if self.verbose > 0:
                    print('Training next epoch...')

            hist = self.model_.fit_generator(
                datagen,
                epochs=self.sub_epochs,
                callbacks=self.callbacks,
                shuffle=shuffle_mode,
                steps_per_epoch=int(np.ceil(X.shape[0] / self.batch_size)),
                verbose=0,
            )

            if early_stopping_monitor.stopped_epoch > 0:
                print('Early stopping.')
                break

            if os.path.exists(self.annoy_index_path):
                os.remove(self.annoy_index_path)

            mean_loss = np.mean(hist.history['loss'][-self.sub_epochs:])
            print('Epoch {}: loss {}'.format(epoch, mean_loss))

            if self.reduce_lr_amount != 1:
                K.set_value(self.model_.optimizer.learning_rate,
                            K.eval(self.model_.optimizer.lr) * self.reduce_lr_amount)
                if self.verbose:
                    print('New learning rate set to: {}'.format(K.eval(self.model_.optimizer.lr)))

        self.loss_history_ += hist.history['loss']

    def fit(self, X, Y, shuffle_mode=True):
        """
        Fit an meezer model.

        Parameters
        ----------
        X: np.array, shape (n_samples, n_features)
            Data to be embedded.
        Y: np.array, shape (n_samples)
            Labels for data in `X`. If `Y` contains -1 labels, and 'sparse_categorical_crossentropy'
            is the loss function, semi-supervised learning will be used.
        shuffle_mode: bool
            Parameter to be sent to Keras `fit_generator` function (default True)

        """
        self._fit(X, Y, shuffle_mode)

        return self

    def fit_transform(self, X, Y=None, shuffle_mode=True):
        """
        Fit to data then transform

        Parameters
        ----------
        X: np.array, shape (n_samples, n_features)
            Data to be embedded.
        Y: np.array, shape (n_samples)
            Labels for data in `X`. If `Y` contains -1 labels, and 'sparse_categorical_crossentropy'
            is the loss function, semi-supervised learning will be used.
        shuffle_mode: bool
            Parameter to be sent to Keras `fit_generator` function (default True)

        Returns
        -------
        X_new : np.array, shape (n_samples, embedding_dims)
            Embedding of the new data in low-dimensional space.

        """
        self.fit(X, Y, shuffle_mode)

        return self.transform(X)

    def transform(self, X, verbose=True):
        """
        Transform X into the existing embedded space and return that
        transformed output.

        Parameters
        ----------
        X: np.array, shape (n_samples, n_features)
            New data to be transformed.
        verbose: bool

        Returns
        -------
        X_new: np.array, shape (n_samples, embedding_dims)
            Embedding of the new data in low-dimensional space.

        """
        embeddings = self.encoder.predict(X, verbose=verbose)

        return embeddings

    def score_samples(self, X):
        """
        Passes X through classification network to obtain predicted supervised values.

        Parameters
        ----------
        X: np.array, shape (n_samples, n_features)
            Data to be passed through classification network.

        Returns
        -------
        X_new: np.array, shape (n_samples, embedding_dims)
            Softmax class probabilities of the data.

        """
        softmax_output = self.supervised_model_.predict(X, verbose=self.verbose)

        return softmax_output

    def save_model(self, folder_path, overwrite=False):
        """
        Save an meezer model

        Parameters
        ----------
        folder_path: string
            Path to serialised model files and metadata
        overwrite: bool
            Whether or not to overwrite existing data (default False)

        Side Effects
        -------------
        Writes model files to disk.

        """
        if overwrite:
            if os.path.exists(folder_path):
                shutil.rmtree(folder_path)
        os.makedirs(folder_path)
        # serialize weights to HDF5
        self.model_.save(os.path.join(folder_path, 'meezer_model.h5'))
        # Have to serialize supervised model separately
        if self.supervised_model_ is not None:
            self.supervised_model_.save(
                os.path.join(folder_path, 'supervised_model.h5')
            )

        json.dump(
            self.__getstate__(),
            open(os.path.join(folder_path, 'meezer_params.json'), 'w'),
        )

    def load_model(self, folder_path):
        """
        Load meezer model

        Parameters
        ----------
        folder_path: string
            Path to serialised model files and metadata

        Returns
        -------
        model: Meezer instance

        """
        meezer_config = json.load(
            open(os.path.join(folder_path, 'meezer_params.json'), 'r')
        )
        self.__dict__ = meezer_config

        loss_function = triplet_loss(self.distance, self.margin)
        self.model_ = load_model(
            os.path.join(folder_path, 'meezer_model.h5'),
            custom_objects={'tf': tf, loss_function.__name__: loss_function},
        )
        self.encoder = self.model_.layers[3]
        self.encoder._make_predict_function()

        # If a supervised model exists, load it
        supervised_path = os.path.join(folder_path, 'supervised_model.h5')
        if os.path.exists(supervised_path):
            self.supervised_model_ = load_model(supervised_path)
        return self
